### PR TITLE
Int2019 49 create automatic payment

### PR DESCRIPTION
**Pull request by ed2298**

### CODE REVIEW

File: ClientApp/viking-vault/src/UserDashboard/AutomaticPaymentList.tsx [hunk 1]
```diff
@@ -2,12 +2,19 @@ import  React  from 'react';
 import {constants} from "../Resources/Constants";
 import './ViewAutomaticPayments.css'
 import ActivatePaymentToggle from "../AdminDashboard/ActivatePaymentToggle"
+import {CreateAutomaticPaymentForm} from "./CreateAutomaticPaymentModal"
+import add_payment from "../Resources/images/add_payment.png"
```
Comment (Slitthe) [inline ClientApp/viking-vault/src/UserDashboard/AutomaticPaymentList.tsx]: Please keep this variable's naming style consistent to the rest of the application, which uses `camelCase`, also I think that naming this `addPaymentImg` would be more descriptive

File: ClientApp/viking-vault/src/UserDashboard/CreateAutomaticPaymentModal.tsx [hunk 2]
```diff
@@ -0,0 +1,258 @@
+import React from "react";
+import { constants } from "../Resources/Constants.js";
+import '../AdminDashboard/AttachCardModal.css';
+
+let regexCheckIfPositiveFloat = "^[-+]?[0-9]*\.?[0-9]+([eE][-+]?[0-9]+)?$";
+let currentDate = new Date();
```
Comment (Slitthe) [inline ClientApp/viking-vault/src/UserDashboard/CreateAutomaticPaymentModal.tsx]: If you don't change these variables, make them constants `const`

File: ClientApp/viking-vault/src/UserDashboard/CreateAutomaticPaymentModal.tsx [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+        else{
+            fetch(constants.baseUrl+"company", {
+                method: "GET",
+                headers: {
+                'Accept': 'application/json',
+                'Content-Type': 'application/json',
+                'x-access-token' : token.toString()
+                }})
+            .then( response => 
+                {
```
Comment (Slitthe) [inline ClientApp/viking-vault/src/UserDashboard/CreateAutomaticPaymentModal.tsx:155]: JS convention: curly bracket on the same line
`blah {`

`}`

File: ClientApp/viking-vault/src/UserDashboard/ViewAutomaticPayments.css [hunk 4]
```diff
@@ -11,4 +11,24 @@
 .automatic-payments-list {
     position: relative;
     right: 4%;
-}
\ No newline at end of file
+}
+
+.add_payment_icon {
```
Comment (Slitthe) [inline ClientApp/viking-vault/src/UserDashboard/ViewAutomaticPayments.css]: Keep the class naming convention consistent, your application uses `kebab-case` for CSS naming, use that here as well

File: Server/VikingVault.Services/AutomaticPaymentService.cs [hunk 5]
```diff
@@ -54,5 +55,29 @@ public List<AutomaticPaymentDTO> GetAllAutomaticPayments(string token)
 
             return automaticPaymentData;
         }
+
+        public AutomaticPayment CreateAutomaticPayment(AutomaticPaymentDTO automaticPaymentDTO, string token)
+        {
+            var tokenObject = new JwtSecurityToken(token);
+            int userId = Int32.Parse(tokenObject.Payload["Id"].ToString());
```
Comment (Slitthe) [inline Server/VikingVault.Services/AutomaticPaymentService.cs]: Use the same data type in the parse as it is in the variable declaration. (I know that `int` and `Int32` are the same but its confusing)

File: ClientApp/viking-vault/src/UserDashboard/AutomaticPaymentList.tsx [hunk 6]
```diff
@@ -1,10 +1,16 @@
...
 const baseUrl = constants.baseUrl + "AutomaticPayment";
 
+interface IModals {
+    openCreateAutomaticPaymentModal: boolean
+}
+
 interface IAutomaticPayment {
+    companyId: number,
     companyName: string,
     amount: Number,
```
Comment (OlahCatalina) [inline ClientApp/viking-vault/src/UserDashboard/AutomaticPaymentList.tsx]: Please use "number" for amount instead of "Number" and eventually investigate the difference between these two.

**Review by OlahCatalina — APPROVED** [review]

File: ClientApp/viking-vault/src/UserDashboard/CreateAutomaticPaymentModal.tsx [hunk 7]
```diff
@@ -0,0 +1,255 @@
...
+        }
+    }
+
+    private initialPaymentDateIsAfterCurrentDate = () => {
+        if(this.state.automaticPayment.initialPaymentDate !== "" && new Date(this.state.automaticPayment.initialPaymentDate) >= currentDate)
+            return  {
+                message: "Ok",
+                class: "alert alert-success"
+            }
+        return  {
```
Comment (crisfsabau) [inline ClientApp/viking-vault/src/UserDashboard/CreateAutomaticPaymentModal.tsx:93]: you have 2 returns in this if, shouldn't the second return be on the else branch?

File: ClientApp/viking-vault/src/UserDashboard/CreateAutomaticPaymentModal.tsx [hunk 8]
```diff
@@ -0,0 +1,255 @@
...
+            companyId: this.getIdOfCompanyByName(this.state.automaticPayment.companyName),
+            companyName: this.state.automaticPayment.companyName,
+            amount: this.state.automaticPayment.amount,
+            initialPaymentDate: new Date(this.state.automaticPayment.initialPaymentDate),
+            lastPaymentDate: new Date(2000,0,2),
+        };
+    }
+
+    getAllCompanies = () => {
+        var companies;
```
Comment (crisfsabau) [inline ClientApp/viking-vault/src/UserDashboard/CreateAutomaticPaymentModal.tsx]: don't use var. Use let or const. let for variables that can be reassigned and const for read only variables

File: ClientApp/viking-vault/src/UserDashboard/CreateAutomaticPaymentModal.tsx [hunk 9]
```diff
@@ -0,0 +1,255 @@
...
+        var token = sessionStorage.getItem("Authentication-Token");
```
Comment (crisfsabau) [inline ClientApp/viking-vault/src/UserDashboard/CreateAutomaticPaymentModal.tsx]: same as above

File: ClientApp/viking-vault/src/UserDashboard/CreateAutomaticPaymentModal.tsx (hunk 3 above)
Comment (crisfsabau) [reply ClientApp/viking-vault/src/UserDashboard/CreateAutomaticPaymentModal.tsx:155]: is this resolved? I don't see it as resolved

File: ClientApp/viking-vault/src/UserDashboard/CreateAutomaticPaymentModal.tsx [hunk 10]
```diff
@@ -0,0 +1,255 @@
...
+        }
+    }
+
+    private amountMustBePositive = () => {
+        if(this.state.automaticPayment.amount.match(regexCheckIfPositiveFloat))
+            return  {
+                message: "Ok",
+                class: "alert alert-success"
+            }
+        return  {
```
Comment (crisfsabau) [inline ClientApp/viking-vault/src/UserDashboard/CreateAutomaticPaymentModal.tsx:80]: Here you have 2 returns in the same branch of the if statement. Did you wanna return after the if, maybe? or in the else branch?

File: Server/VikingVault.Services/CompanyService.cs [hunk 11]
```diff
@@ -73,9 +73,9 @@ public List<CompanyDataDTO> GetAllCompanies()
                 var companies = _dbContext.User.Where(user => user.Role.Type == RoleEnum.Company.ToString()).ToList();
                 foreach (var company in companies)
                 {
-                    var account = _dbContext.BankAccount
+                    var account = _dbContext.BankAccount.Include(bankAccount => bankAccount.User)
                         .SingleOrDefault(bankAccount => bankAccount.User.Id == company.Id &&
-                                                        bankAccount.CurrencyType == CurrencyEnum.Ron.ToString());
+                                                        bankAccount.CurrencyType.Equals("Ron"));
```
Comment (crisfsabau) [inline Server/VikingVault.Services/CompanyService.cs:78]: why did you change this?

File: ClientApp/viking-vault/src/UserDashboard/CreateAutomaticPaymentModal.tsx (hunk 7 above)
Comment (ed2298) [reply ClientApp/viking-vault/src/UserDashboard/CreateAutomaticPaymentModal.tsx:93]: Not necessarily. If cond=true, then the return inside the if statement is returned, which means that the latter return won't be reached. Otherwise, the code inside the if statement is not executed, so the second return is "executed".

Comment (crisfsabau) [reply ClientApp/viking-vault/src/UserDashboard/CreateAutomaticPaymentModal.tsx:93]: yes but I don't see the closing brace of the if before the return

File: Server/VikingVault.Services/CompanyService.cs (hunk 11 above)
Comment (ed2298) [reply Server/VikingVault.Services/CompanyService.cs:78]: I think it is safer to use equals for strings, instead of == operator.

File: ClientApp/viking-vault/src/UserDashboard/CreateAutomaticPaymentModal.tsx (hunk 7 above)
Comment (ed2298) [reply ClientApp/viking-vault/src/UserDashboard/CreateAutomaticPaymentModal.tsx:93]: That is because there is no opening brace ;)

Comment (crisfsabau) [reply ClientApp/viking-vault/src/UserDashboard/CreateAutomaticPaymentModal.tsx:93]: Then use braces, to be more clear

Comment (ed2298) [reply ClientApp/viking-vault/src/UserDashboard/CreateAutomaticPaymentModal.tsx:93]: I will enclose the code from the if statement in curly brackets
